File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
class PreInBinaryTree{
    constructor(){
        this.root = null;
    }

    constructTree(preorder, preStart, preEnd, inStart, inEnd, inHashmap){
        // Base Case:
        // when there's no nodes between preStart & preEnd or inStart & inEnd;
        // thst's actually empty array;
        if(preStart > preEnd || inStart > inEnd){
            return null;
        }

        const nodeVal = preorder[preStart];

        const root = new TreeNode(nodeVal);

        // Getting the root node index from inorder traversal array
        // to determine the left subTree and right subTree;

        const inRoot = inHashmap.get(nodeVal);
        const nodesCount = inRoot - inStart;

        const leftPreEnd = preStart + nodesCount;

        root.left = this.constructTree(preorder, preStart + 1, leftPreEnd, inStart, inRoot - 1, inHashmap);
        root.right = this.constructTree(preorder, leftPreEnd + 1, preEnd, inRoot + 1, inEnd, inHashmap);

        return root;
    }
}
var buildTree = function(preorder, inorder) {
    // Logic:
    // Making use of preorder properties and inorder properties:
    // preorder: Root -> Left -> Right
    // inorder: Left -> Root -> Right
    const inOrderElsHashmap = new Map();

    for(let idxI = 0; idxI < inorder.length; idxI++){
        const nodeEl = inorder[idxI];
        inOrderElsHashmap.set(nodeEl, idxI);
    }

    const constructPreInBiTree = new PreInBinaryTree();

    const root = constructPreInBiTree.constructTree(preorder, 0, preorder.length - 1, 0, inorder.length - 1, inOrderElsHashmap);

    return root;
};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class PreInBinaryTree():
    def __init__(self):
        self.root = None

    def construct_tree(self, preorder, pre_strt, pre_end, in_strt, in_end, in_hashmap):

        # Base Case:
        if pre_strt > pre_end or in_strt > in_end:
            return None

        node_val = preorder[pre_strt]

        root = TreeNode(node_val)

        in_root = in_hashmap[node_val]
        nodes_count = in_root - in_strt

        left_pre_end = pre_strt + nodes_count

        root.left = self.construct_tree(preorder, pre_strt + 1, left_pre_end, in_strt, in_root - 1, in_hashmap)

        root.right = self.construct_tree(preorder, left_pre_end + 1, pre_end, in_root + 1, in_end, in_hashmap)

        return root

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:

        inorder_els_map = {}

        for idx in range(len(inorder)):
            inorder_els_map[inorder[idx]] = idx
        
        construct_pre_in_bitree = PreInBinaryTree()

        root = construct_pre_in_bitree.construct_tree(preorder, 0, len(preorder) - 1, 0, len(inorder) - 1, inorder_els_map)

        return root
        